src/Interpreters/RedundantFunctionsInOrderByVisitor.h

```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include <Interpreters/InDepthNodeVisitor.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTOrderByElement.h>
+#include <Parsers/ASTSelectQuery.h>
+
+namespace DB
+{
+
+class RedundantFunctionsInOrderByMatcher
+{
+public:
+    struct Data
+    {
+        std::unordered_set<String> & keys;
+        bool should_be_erased = false;
+        bool done = false;
+    };
+
+    static void visit(const ASTPtr & ast, Data & data)
+    {
+        auto * ast_function = ast->as<ASTFunction>();
+        if (ast_function)
+        {
+            visit(*ast_function, data);
+        }
+    }
+
+    static void visit(ASTFunction & ast_function, Data & data)
+    {
+        if (data.done)
+            return;
+
+        auto arguments = ast_function.arguments;
+
+        if (arguments->children.size() != 1)
```

Comment:
Most of functions are deterministic even if they have multiple arguments. I.e. `x + y`. It's very strong restriction to optimize only functions of 1 argument. We need optimize more :)